### PR TITLE
Collision detection with redis lock

### PR DESCRIPTION
**Pull request by whyang9701**
透過 redis.set 的 NX flag 機制實現玩家不可出現在同一格

### CODE REVIEW

File: services/gateway/gateway-service.mjs [hunk 1]
```diff
@@ -272,26 +277,40 @@ class GatewayService {
...
 
-  updatePosition(uid,x,y,facing){
-    assert.fail('Not implemented');
+  _getDistanceSquare(a,b){
+    return Math.pow(Math.abs(a.x - b.x),2) + Math.pow(Math.abs(a.y - b.y),2);
   }
 
-  getDistanceSquare(a,b){
-    return Math.abs(a.x - b.x)^2 + Math.abs(a.y - b.y)^2
+  _borderCheck(msg,mapSize,mapShift){
```
Comment (john0312) [inline services/gateway/gateway-service.mjs:292]: Why do you need mapshift?

File: services/gateway/gateway-service.mjs [hunk 2]
```diff
@@ -272,26 +277,40 @@ class GatewayService {
...
+    let targetPosition = {x:msg.x , y:msg.y};
+    let distanceSquire = this._getDistanceSquare(lastPosition,targetPosition);
+    let speed = 1 ;
+    if(distanceSquire > speed){
+      return false;
+    } 
+    return true;
   }
+
+  _
```
Comment (john0312) [inline services/gateway/gateway-service.mjs:313]: What's this?

File: services/gateway/gateway-service.mjs [hunk 3]
```diff
@@ -272,26 +277,40 @@ class GatewayService {
...
+    return true;
   }
 
-  checkPositionEmpty(x,y){
-    return gameMap.getCell(x,y);
+  _nearByGridCheck(msg,lastRecord){
+    let lastPosition = {x:lastRecord.x , y:lastRecord.y};
+    let targetPosition = {x:msg.x , y:msg.y};
+    let distanceSquire = this._getDistanceSquare(lastPosition,targetPosition);
+    let speed = 1 ;
```
Comment (john0312) [inline services/gateway/gateway-service.mjs:306]: If we are dealing with speed, we need to take the time (duration between 2 moves) into consideration as well.

File: services/gateway/gateway-service.mjs [hunk 4]
```diff
@@ -216,45 +218,48 @@ class GatewayService {
...
-    let positionB = {x:msg.x,y:msg.y};
-    let distanceSquire = this.getDistanceSquare(positionA,positionB);
-
-    //overspeed , are you fly?
-    if(distanceSquire > speed ^ 2){
-      this.broadcastResetUser(socket,uid,positionA.x,positionA.y,positionA.facing)
-      return
+    let mapShift = {x: 6 , y: 3};
+    let mapSize = this.gameMap.getMapSize();
+    let lastRecord = await this.broadcaster.gameState.getPlayer(socket.playerID);
```
Comment (john0312) [inline services/gateway/gateway-service.mjs:229]: We might have a race condition here because broadcasting takes time.

**Comment by whyang9701** [comment]
可以麻煩John 有空的時候再幫我看看嗎

File: services/gateway/gateway-service.mjs [hunk 5]
```diff
@@ -163,11 +163,16 @@ class GatewayService {
     }
 
     // Synchronize the state.
-    let firstLocation = {playerID: playerID, displayName:
-      socket.playerData.displayName};
-    firstLocation.mapCoord = socket.playerData.mapCoord;
+    let firstLocation = {playerID: playerID, 
+      displayName: socket.playerData.displayName};
+    firstLocation.x = socket.playerData.x;
```
Comment (john0312) [inline services/gateway/gateway-service.mjs:168]: What about which world it is?

File: services/gateway/gateway-service.mjs [hunk 6]
```diff
@@ -163,11 +163,16 @@ class GatewayService {
...
+    firstLocation.y = socket.playerData.y;
     firstLocation.facing = 'D';
     firstLocation.displayChar = socket.playerData.displayChar;
+    socket.playerData.lastMovingTime = Date.now();
+    //try occupy grid
+    await this.dir.redis.setAsync(
+      [`${firstLocation.x}-${firstLocation.y}`, 'occupied', 'NX']);
```
Comment (john0312) [inline services/gateway/gateway-service.mjs:175]: You might need to include which world it is.

File: services/gateway/gateway-service.mjs [hunk 7]
```diff
@@ -227,36 +236,46 @@ class GatewayService {
...
-
-    let positionA = this.getLastPosition(uid)
-    let positionB = {x:msg.x,y:msg.y};
-    let distanceSquire = this.getDistanceSquare(positionA,positionB);
-
-    //overspeed , are you fly?
-    if(distanceSquire > speed ^ 2){
-      this.broadcastResetUser(socket,uid,positionA.x,positionA.y,positionA.facing)
-      return
+    let mapSize = this.gameMap.getMapSize();
```
Comment (john0312) [inline services/gateway/gateway-service.mjs:239]: You might need to specify which world it is.

File: services/gateway/gateway-service.mjs [hunk 8]
```diff
@@ -275,25 +294,39 @@ class GatewayService {
     return true;
   }
 
-  async broadcastResetUser(socket, uid, mapCoord, facing) {
-    socket.broadcast.emit('location', {uid, mapCoord, facing});
+  _getDistanceSquare(a,b){
+    return Math.pow(Math.abs(a.x - b.x),2) + Math.pow(Math.abs(a.y - b.y),2);
```
Comment (john0312) [inline services/gateway/gateway-service.mjs:298]: Consider using Manhattan distance so we can have integer distance?

File: services/gateway/gateway-service.mjs [hunk 9]
```diff
@@ -8,8 +8,7 @@ const require = createRequire(import.meta.url);
 import assert from 'assert';
 const {Server} = require('socket.io');
 const config = require('config');
-
-import { get } from 'http';
+const moveingRequestThreshold = config.get('movingRequestThreshold')
```
Comment (john0312) [inline services/gateway/gateway-service.mjs:11]: "moving"

File: services/gateway/gateway-service.mjs [hunk 10]
```diff
@@ -275,25 +294,39 @@ class GatewayService {
     return true;
   }
 
-  async broadcastResetUser(socket, uid, mapCoord, facing) {
-    socket.broadcast.emit('location', {uid, mapCoord, facing});
+  _getDistanceSquare(a,b){
```
Comment (john0312) [inline services/gateway/gateway-service.mjs:297]: [Not required in this PR] Please consider moving the checks below into a library that can be used both in backend and frontend.

**Pull request merged**